src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -0,0 +1,297 @@
+//! Make use of a K-bit lookup table to decompose a field element into K-bit
+//! words.
+
+use crate::primitives::sinsemilla::lebs2ip_k;
+use halo2::{
+    circuit::{Layouter, Region},
+    plonk::{Advice, Column, ConstraintSystem, Error, Fixed, Permutation},
+    poly::Rotation,
+};
+use std::marker::PhantomData;
+
+use ff::PrimeFieldBits;
+
+use super::*;
+
+#[derive(Debug, Clone)]
+pub struct LookupRangeCheckConfig<F: FieldExt + PrimeFieldBits, const K: usize> {
+    q_lookup: Column<Fixed>, // This is passed in.
+    running_sum: Column<Advice>,
+    table_idx: Column<Fixed>,
+    perm: Permutation,
+    _marker: PhantomData<F>,
+}
+
+impl<F: FieldExt + PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K> {
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_lookup: Column<Fixed>,
+        running_sum: Column<Advice>,
+        table_idx: Column<Fixed>,
+        perm: Permutation,
+    ) -> Self {
+        let config = LookupRangeCheckConfig {
+            q_lookup,
+            running_sum,
+            table_idx,
+            perm,
+            _marker: PhantomData,
+        };
+
+        meta.lookup(|meta| {
+            let q_lookup = meta.query_fixed(config.q_lookup, Rotation::cur());
+            let z_cur = meta.query_advice(config.running_sum, Rotation::cur());
+            let z_next = meta.query_advice(config.running_sum, Rotation::next());
+            //    z_i = 2^{K}⋅z_{i + 1} + a_i
+            // => a_i = z_i - 2^{K}⋅z_{i + 1}
+            let word = z_cur - z_next * F::from_u64(1 << K);
+            let table = meta.query_fixed(config.table_idx, Rotation::cur());
+
+            vec![(q_lookup * word, table)]
+        });
+
+        config
+    }
+
+    pub fn load(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
```

Comment:
This lookup table is provided by the Sinsemilla chip in the context of Orchard, but may not be in general.

```suggestion
    #[cfg(test)]
    fn load(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
```